api/mappers/engine.py

```diff
@@ -0,0 +1,404 @@
+) -> Optional[IntegrationModel]:
+    if not integration:
+        return None
+    return IntegrationModel(
+        api_key=integration.api_key,
+        base_url=integration.base_url,
+        entity_selector=integration.entity_selector,
+    )
+
+
+def map_webhook_config_to_engine(
+    webhook_config: Optional["WebhookConfiguration"],
+) -> Optional[WebhookModel]:
+    if not webhook_config:
+        return None
+    return WebhookModel(
+        url=webhook_config.url,
+        secret=webhook_config.secret,
+    )
+
+
+def map_feature_state_to_engine(
+    feature_state: "FeatureState",
+    mv_fs_values: Iterable["MultivariateFeatureStateValue"],
+) -> FeatureStateModel:
+    feature = feature_state.feature
+    feature_segment: Optional["FeatureSegment"] = feature_state.feature_segment
+
+    if feature_segment:
+        feature_segment_model = FeatureSegmentModel(
+            priority=feature_segment.priority,
+        )
+    else:
+        feature_segment_model = None
+
+    return FeatureStateModel(
+        enabled=feature_state.enabled,
+        django_id=feature_state.pk,
+        feature_state_value=feature_state.get_feature_state_value(),
+        featurestate_uuid=feature_state.uuid,
+        feature_segment=feature_segment_model,
+        feature=map_feature_to_engine(feature),
+        multivariate_feature_state_values=[
+            map_mv_fs_value_to_engine(mv_fs_value) for mv_fs_value in mv_fs_values
+        ],
+    )
+
+
+def map_mv_fs_value_to_engine(
+    mv_fs_value: "MultivariateFeatureStateValue",
+) -> MultivariateFeatureStateValueModel:
+    mv_feature_option: "MultivariateFeatureOption" = (
+        mv_fs_value.multivariate_feature_option
+    )
+
+    return MultivariateFeatureStateValueModel(
+        percentage_allocation=mv_fs_value.percentage_allocation,
+        id=mv_fs_value.id,
+        mv_fs_value_uuid=mv_fs_value.uuid,
+        multivariate_feature_option=map_mv_option_to_engine(mv_feature_option),
+    )
+
+
+def map_feature_to_engine(feature: "Feature") -> FeatureModel:
+    return FeatureModel(id=feature.pk, name=feature.name, type=feature.type)
+
+
+def map_mv_option_to_engine(
+    mv_option: "MultivariateFeatureOption",
+) -> MultivariateFeatureOptionModel:
+    return MultivariateFeatureOptionModel(value=mv_option.value, id=mv_option.id)
+
+
+def map_environment_to_engine(
+    environment: "Environment",
+) -> EnvironmentModel:
+    """
+    Maps Core API's `environments.models.Environment` mode instance to the
+    flag_engine environment document.
+    Before building the document, takes care of resolving relationships and
+    feature versions.
+
+    :param Environment environment: the environment to map
+    :rtype EnvironmentModel
+    """
+    project: "Project" = environment.project
+    organisation: "Organisation" = project.organisation
+
+    # Read relationships.
+    project_segments: List["Segment"] = project.segments.all()
+    project_segment_rules_by_segment_id: Dict[
```

Comment:
I don't quite follow why we need this dictionary?

Reply:
I guess we could annotate segment instances with `SegmentRule` lists, but I decided it's a bit cleaner to store those separately.